spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -736,7 +769,14 @@ public void update(InternalRow meta, InternalRow id, InternalRow row) throws IOE
 
     @Override
     public void insert(InternalRow row) throws IOException {
-      delegate.insert(row, dataSpec, null);
+      reinsert(null, row);
+    }
+
+    @Override
+    public void reinsert(InternalRow meta, InternalRow row) throws IOException {
```

Comment:
I looked at `DeltaWithMetadataWritingSparkTask` in Spark code. it seems does what I suggested in another comment regarding using projection pattern.
```
        case UPDATE_OPERATION =>
          rowProjection.project(row)
          rowIdProjection.project(row)
          metadataProjection.project(row)
          writer.update(metadataProjection, rowIdProjection, rowProjection)
```

Reply:
can probably leverage/extends from the `ProjectingInternalRow` class in Spark